Move some distributed C++ GTest unit tests from test.sh to Python wrapper

haovic opened the pull request:
Summary:
This is the beginning of transitioning away from test.sh: https://github.com/pytorch/pytorch/issues/64067

Moved execution of FileStoreTest, HashStoreTest, TCPStoreTest from test.sh to run_test.py. Since these are compiled tests, the logic of test.sh was just migrated and the tests are run through subprocess.

Test Plan:
```
python3 test/run_test.py --include cpp/c10d/FileStoreTest cpp/c10d/HashStoreTest cpp/c10d/TCPStoreTest
```

Wait for CI

Differential Revision: D30662844

malfet commented on test/run_test.py:
```diff
@@ -25,6 +26,125 @@
+            print_to_stderr("Using build folder")
+            TEST_BASE_DIR = BUILD_BIN_DIR
+
+        SUDO = "sudo" if "ppc64le" in BUILD_ENVIRONMENT else ""
+        shell((SUDO + " ln -sf " + TORCH_LIB_DIR + "/libc10* " + TEST_BASE_DIR).split())
+        shell((SUDO + " ln -sf " + TORCH_LIB_DIR + "/libcaffe2* " + TEST_BASE_DIR).split())
+        shell((SUDO + " ln -sf " + TORCH_LIB_DIR + "/libmkldnn* " + TEST_BASE_DIR).split())
+        shell((SUDO + " ln -sf " + TORCH_LIB_DIR + "/libnccl* " + TEST_BASE_DIR).split())
+        shell((SUDO + " ln -sf " + TORCH_LIB_DIR + "/libtorch* " + TEST_BASE_DIR).split())
```
Please don't do that, but instead just use `LD_LIBRARY_PATH`
```suggestion
     os.env["LD_LIBRARY_PATH"] = f"{TORCH_LIB_DIR};{os.env.get('LD_LIBRARY_PATH',''}"
```

malfet commented on test/run_test.py:
```diff
@@ -25,6 +26,125 @@
 from typing import Dict, Optional, List
 
 REPO_ROOT = pathlib.Path(__file__).resolve().parent.parent
+TORCH_INSTALL_DIR = site.getsitepackages()[0] + "/torch"
+TORCH_BIN_DIR = TORCH_INSTALL_DIR + "/bin"
+TORCH_LIB_DIR = TORCH_INSTALL_DIR + "/lib"
+TORCH_TEST_DIR = TORCH_INSTALL_DIR + "/test"
```
Please avoid string concat for path manupulation, but instead either use a `Path` class or `os.path.join`
```suggestion
TORCH_INSTALL_DIR = os.path.join(site.getsitepackages()[0], "torch")
TORCH_BIN_DIR = os.path.join(TORCH_INSTALL_DIR, "bin")
TORCH_LIB_DIR = os.path.join(TORCH_INSTALL_DIR, "lib")
TORCH_TEST_DIR = os.path.join(TORCH_INSTALL_DIR, "test")
```

malfet commented on test/run_test.py:
```diff
@@ -922,6 +1051,33 @@ def run_test_module(test: str, test_directory: str, options) -> Optional[str]:
+            return None
+
+        test_reports_dir = "test/test-reports/cpp-distributed/test_distributed"
+        shell(("mkdir -p " + test_reports_dir).split())
```
This is very prone to errors (especially if test_reports_dir contains spaces)
```suggestion
        os.makedirs(test_reports_dir, exists_ok=True)
```

haovic committed:
Move some distributed C++ GTest unit tests from test.sh to Python wrapper (#64252)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/64252

This is the beginning of transitioning away from test.sh: https://github.com/pytorch/pytorch/issues/64067

Moved execution of FileStoreTest, HashStoreTest, TCPStoreTest from test.sh to run_test.py. Since these are compiled tests, the logic of test.sh was just migrated and the tests are run through subprocess.

Test Plan:
```
python3 test/run_test.py --include cpp/c10d/FileStoreTest cpp/c10d/HashStoreTest cpp/c10d/TCPStoreTest
```

Wait for CI

Differential Revision: D30662844

fbshipit-source-id: 3807399df05a0f5a08e06f11eba68f602b2bf014
